freqtrade/data/dataprovider.py

```diff
@@ -150,6 +151,80 @@ def _add_external_df(
         self.__producer_pairs_df[producer_name][pair_key] = (dataframe, _last_analyzed)
         logger.debug(f"External DataFrame for {pair_key} from {producer_name} added.")
 
+    def _add_external_candle(
+        self,
+        pair: str,
+        dataframe: DataFrame,
+        last_analyzed: datetime,
+        timeframe: str,
+        candle_type: CandleType,
+        producer_name: str = "default"
+    ) -> Tuple[bool, int]:
+        """
+        Append a candle to the existing external dataframe. The incoming dataframe
+        must have at least 1 candle.
+
+        :param pair: pair to get the data for
+        :param timeframe: Timeframe to get data for
+        :param candle_type: Any of the enum CandleType (must match trading mode!)
+        :returns: False if the candle could not be appended, or the int number of missing candles.
+        """
+        pair_key = (pair, timeframe, candle_type)
+
+        if dataframe.empty:
+            # The incoming dataframe must have at least 1 candle
+            return (False, 0)
+
+        if (producer_name not in self.__producer_pairs_df
+           or pair_key not in self.__producer_pairs_df[producer_name]):
+            # We don't have data from this producer yet,
+            # or we don't have data for this pair_key
+            # return False and 1000 for the full df
+            return (False, 1000)
+
+        existing_df, _ = self.__producer_pairs_df[producer_name][pair_key]
+
+        # Handle overlapping candles
+        old_candles = existing_df[
+            ~existing_df['date'].isin(
+                dataframe['date']
+            )
+        ]
+        overlapping_candles = existing_df[
+            existing_df['date'].isin(
+                dataframe['date']
+            )
+        ]
+        new_candles = dataframe[
+            ~dataframe['date'].isin(
+                existing_df['date']
+            )
+        ]
+
+        if overlapping_candles:
+            existing_df = concat([old_candles, overlapping_candles], axis=0)
```

Comment:
i think instead of filtering 3 times (and then concatting twice, once here and once in the other method) - it would be much simpler (and safer to some degree) to do something like this:

```
# Drop all duplicate candles
existing_df1 = existing_df[existing_df['date'] < incomming_first]
...
```

If the above modifies the existing dataframe - then there was an overlapping candle, so we technically can never get a candle difference.
if we get a difference - this also automatically implies that the dataframe was not modified - then we obviously return and request more.

After that, a simple concat will suffice (we removed all duplicates first, so we won't create duplicates) (obviously followed by a limit to the 1500 candles).

---

the current method has the downside that, in theory, it could drop canldes from the "middle". 
We assume that ain't going to happen as we control the producer - but we don't really - so with "in" filtering, we could end up with an unsorted dataframe (if the incomming dataframe itself had leaks).
It's also 2 more pandas operations than necessary.

Reply:
I've implemented the changes i mentioned above - as the current method was failing with several errors (just checkout 5811363 to see these errors).

based on the tests - i think the current method will work just fine - though i'm not sure if this shouldn't be the only entrypoint for emc.
we do have "is full dataframe" handling here, too - so it's kinda redundant to move this decision to emc.

Reply:
it's a thought so far, nothing i investigated deeper.

in the end, we don't need 2 entry-points to add an external dataframe. 
we can take the decision "is this a complete one" - or an "aggregatable one" within this function - which will simplify the logic a lot (right now we do this decision twice, once in EMC to call `add_external_df()` - and once within this method - if the resulting, truncated dataframe was empty.

Reply:
I'm not sure I agree. To me, the logic for deciding what to do with messages should stay in the EMC.

Reply:
then this whole function is misplaced where it is now.

Part of this decision is already here - but we now take this decision twice.